### ISSUE TITLE
Using type_convert() results in a corrupt data frame

### ISSUE THREAD

**Issue by isaactpetersen** [tidyverse/readr#1016]
Hi, I receive a warning when using `type_convert()`:

```r
type_convert(df)

Warning message:
In format.data.frame(if (omit) x[seq_len(n0), , drop = FALSE] else x,  :
  corrupt data frame: columns will be truncated or padded with NAs
```

I'm not sure how to provide a reproducible example because I don't receive the warning on other data files.  My data file is 1,973 rows and 5,329 columns.  Here's my session info:

```r
> sessionInfo()
R version 3.5.3 (2019-03-11)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 14393)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] graphics  grDevices datasets  stats     utils     methods   base     

other attached packages:
 [1] digest_0.6.20       lubridate_1.7.4     REDCapR_0.9.10.9002 xtable_1.8-4        forcats_0.4.0       dplyr_0.8.3         purrr_0.3.2         readr_1.3.1        
 [9] tidyr_0.8.3         tibble_2.1.3        tidyverse_1.2.1     nlme_3.1-137        psych_1.8.12        mix_1.0-10          mitools_2.4         stringr_1.4.0      
[17] Hmisc_4.2-0         ggplot2_3.2.0       Formula_1.2-3       survival_2.44-1.1   lattice_0.20-38    

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.1          packrat_0.5.0       assertthat_0.2.1    zeallot_0.1.0       cellranger_1.1.0    R6_2.4.0            backports_1.1.4     acepack_1.4.1      
 [9] httr_1.4.0          pillar_1.4.2        rlang_0.4.0         curl_4.0            readxl_1.3.1        lazyeval_0.2.2      rstudioapi_0.10     data.table_1.12.2  
[17] rpart_4.1-15        Matrix_1.2-17       checkmate_1.9.4     splines_3.5.3       foreign_0.8-71      htmlwidgets_1.3     munsell_0.5.0       broom_0.5.2        
[25] compiler_3.5.3      modelr_0.1.4        xfun_0.8            pkgconfig_2.0.2     base64enc_0.1-3     mnormt_1.5-5        htmltools_0.3.6     nnet_7.3-12        
[33] tidyselect_0.2.5    gridExtra_2.3       htmlTable_1.13.1    crayon_1.3.4        withr_2.1.2         grid_3.5.3          jsonlite_1.6        gtable_0.3.0       
[41] DBI_1.0.0           magrittr_1.5        scales_1.0.0        cli_1.1.0           stringi_1.4.3       latticeExtra_0.6-28 xml2_1.2.0          fortunes_1.5-4     
[49] generics_0.0.2      vctrs_0.2.0         RColorBrewer_1.1-2  tools_3.5.3         glue_1.3.1          hms_0.5.0           parallel_3.5.3      colorspace_1.4-1   
[57] cluster_2.1.0       rvest_0.3.4         knitr_1.23          haven_2.1.1   
```

**Comment by jimhester** [comment]
Sorry, this is not enough information for me to do anything actionable with. Unless you can provide some example data (perhaps a subset of your original dataset?) that exhibits the issue I will have to close this.

**Comment by isaactpetersen** [comment]
Okay, I'm attaching a subset of the data file.  When I run `type_convert()` on the subset, I get the warning.  It appears to be something related to the `proc_datelabvisit2` variable.
[test1.zip](https://github.com/tidyverse/readr/files/3438310/test1.zip)

**Comment by cderv** [comment]
Hi, 
I tried to investigate this issue and stumble upon something strange. here is the step of my investigation. 

TL;DR

* The issue comes from some `col_time()` column converted to `hms` class
* The issue comes also from some strange behavior because of the big number of column

## Steps

This is the code I use to import the data. 
``` r
# download file
url <- "https://github.com/tidyverse/readr/files/3438310/test1.zip"
local_file <- tempfile(fileext = ".zip")
download.file(url, local_file, mode = "wb")
# load Rdata file
load(unz(local_file, filename = "test1.Rdata"))
# the data is called test1
dim(test1)
#> [1]    1 1001
```

You got a table with a lot of columns. After that, reprex is not easy (to show errors...)

By using this code, 
```r
options(warn = 2)
res <- readr::type_convert(test1)
res
traceback()
```
we guess that it comes from `format.data.frame` and indeed we found the warning in the code:
https://github.com/wch/r-source/blob/c1259c418eab8cb1604b32b6f6f96d2b6b5bf508/src/library/base/R/format.R#L248

By going step by step in debug mode, we understand that there is something off with the format step with two issue not related I thing. 
First one, is that the data has one observation, but `omit` gets to TRUE and the data.frame is truncated as if we had more than `getOption("max.print")` lines. This is strange but we'll come back to that. 

Second one if that the warning is thrown because of this line 
https://github.com/wch/r-source/blob/780021752eb83a71e2198019acf069ba8741103b/src/library/base/R/format.R#L224

And while digging around I found this is due to some time column that are convert to class `hms()`.  (column 19 and 21). If we remove them it is working without warning. 
```r
# no warning
readr::type_convert(test1[, -c(19, 21)])
```

If we try only with those it is ok
```r
readr::type_convert(test1[, c(19, 21)])
```

but I think this is because of the interaction with strange behavior I was mentioning. It only happens with a certain number of column. And we can already see that something is odd while printing `test1` (so before `type_convert`) because we see this message
```r
[ reached 'max' / getOption("max.print") -- omitted 1 rows ]
```
very strange for a one row table to not get all its rows printed. 

We can confirm the warning comes back when we use more that `getOption("max.print")`

Try this 
```r
# this works (col 19, 21 are included) without warning and print one row
readr::type_convert(test1[, 1:100])
# this prints no row and we get the warning
readr::type_convert(test1[, 1:101])

So **Why is this happening ?**

We come back to first issue I mentioned. There seems to be something odd in R source code for `print.data.frame` when trying to see if some rows must be omitted
```r
	if(is.null(max)) max <- getOption("max.print", 99999L)
        if(!is.finite(max)) stop("invalid 'max' / getOption(\"max.print\"): ", max)
	## format.<*>() : avoiding picking up e.g. format.AsIs
	omit <- (n0 <- max %/% length(x)) < n
	m <- as.matrix(
	    format.data.frame(if(omit) x[seq_len(n0), , drop=FALSE] else x,
			      digits = digits, na.encode = FALSE))
```
[source](https://github.com/wch/r-source/blob/635a211ecc1e695ad97085d1171cfa4e3067c929/src/library/base/R/dataframe.R#L1512)

This line 
```r
omit <- (n0 <- max %/% length(x)) < n
```
is strange. `length(x)` gives us columns number and it is compared to lines. This changes was introduced in last R versions.

We can reproduce with iris
``` r
options(max.print = nrow(iris) + 100L)
getOption('max.print')
#> [1] 250
iris[1, rep_len(seq_len(ncol(iris)), 250)]
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species Sepal.Length.1
#> 1          5.1         3.5          1.4         0.2  setosa            5.1
#>   Sepal.Width.1 Petal.Length.1 Petal.Width.1 Species.1 Sepal.Length.2
#> 1           3.5            1.4           0.2    setosa            5.1
#>   Sepal.Width.2 Petal.Length.2 Petal.Width.2 Species.2 Sepal.Length.3
#> 1           3.5            1.4           0.2    setosa            5.1
#>   Sepal.Width.3 Petal.Length.3 Petal.Width.3 Species.3 Sepal.Length.4
#> 1           3.5            1.4           0.2    setosa            5.1
#>   Sepal.Width.4 Petal.Length.4 Petal.Width.4 Species.4 Sepal.Length.5
#> 1           3.5            1.4           0.2    setosa            5.1
#>   Sepal.Width.5 Petal.Length.5 Petal.Width.5 Species.5 Sepal.Length.6
#> 1           3.5            1.4           0.2    setosa            5.1
#>   Sepal.Width.6 Petal.Length.6 Petal.Width.6 Species.6 Sepal.Length.7
#> 1           3.5            1.4           0.2    setosa            5.1
#>   Sepal.Width.7 Petal.Length.7 Petal.Width.7 Species.7 Sepal.Length.8
#> 1           3.5            1.4           0.2    setosa            5.1
#>   Sepal.Width.8 Petal.Length.8 Petal.Width.8 Species.8 Sepal.Length.9
#> 1           3.5            1.4           0.2    setosa            5.1
#>   Sepal.Width.9 Petal.Length.9 Petal.Width.9 Species.9 Sepal.Length.10
#> 1           3.5            1.4           0.2    setosa             5.1
#>   Sepal.Width.10 Petal.Length.10 Petal.Width.10 Species.10 Sepal.Length.11
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.11 Petal.Length.11 Petal.Width.11 Species.11 Sepal.Length.12
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.12 Petal.Length.12 Petal.Width.12 Species.12 Sepal.Length.13
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.13 Petal.Length.13 Petal.Width.13 Species.13 Sepal.Length.14
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.14 Petal.Length.14 Petal.Width.14 Species.14 Sepal.Length.15
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.15 Petal.Length.15 Petal.Width.15 Species.15 Sepal.Length.16
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.16 Petal.Length.16 Petal.Width.16 Species.16 Sepal.Length.17
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.17 Petal.Length.17 Petal.Width.17 Species.17 Sepal.Length.18
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.18 Petal.Length.18 Petal.Width.18 Species.18 Sepal.Length.19
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.19 Petal.Length.19 Petal.Width.19 Species.19 Sepal.Length.20
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.20 Petal.Length.20 Petal.Width.20 Species.20 Sepal.Length.21
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.21 Petal.Length.21 Petal.Width.21 Species.21 Sepal.Length.22
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.22 Petal.Length.22 Petal.Width.22 Species.22 Sepal.Length.23
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.23 Petal.Length.23 Petal.Width.23 Species.23 Sepal.Length.24
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.24 Petal.Length.24 Petal.Width.24 Species.24 Sepal.Length.25
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.25 Petal.Length.25 Petal.Width.25 Species.25 Sepal.Length.26
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.26 Petal.Length.26 Petal.Width.26 Species.26 Sepal.Length.27
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.27 Petal.Length.27 Petal.Width.27 Species.27 Sepal.Length.28
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.28 Petal.Length.28 Petal.Width.28 Species.28 Sepal.Length.29
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.29 Petal.Length.29 Petal.Width.29 Species.29 Sepal.Length.30
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.30 Petal.Length.30 Petal.Width.30 Species.30 Sepal.Length.31
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.31 Petal.Length.31 Petal.Width.31 Species.31 Sepal.Length.32
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.32 Petal.Length.32 Petal.Width.32 Species.32 Sepal.Length.33
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.33 Petal.Length.33 Petal.Width.33 Species.33 Sepal.Length.34
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.34 Petal.Length.34 Petal.Width.34 Species.34 Sepal.Length.35
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.35 Petal.Length.35 Petal.Width.35 Species.35 Sepal.Length.36
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.36 Petal.Length.36 Petal.Width.36 Species.36 Sepal.Length.37
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.37 Petal.Length.37 Petal.Width.37 Species.37 Sepal.Length.38
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.38 Petal.Length.38 Petal.Width.38 Species.38 Sepal.Length.39
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.39 Petal.Length.39 Petal.Width.39 Species.39 Sepal.Length.40
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.40 Petal.Length.40 Petal.Width.40 Species.40 Sepal.Length.41
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.41 Petal.Length.41 Petal.Width.41 Species.41 Sepal.Length.42
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.42 Petal.Length.42 Petal.Width.42 Species.42 Sepal.Length.43
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.43 Petal.Length.43 Petal.Width.43 Species.43 Sepal.Length.44
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.44 Petal.Length.44 Petal.Width.44 Species.44 Sepal.Length.45
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.45 Petal.Length.45 Petal.Width.45 Species.45 Sepal.Length.46
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.46 Petal.Length.46 Petal.Width.46 Species.46 Sepal.Length.47
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.47 Petal.Length.47 Petal.Width.47 Species.47 Sepal.Length.48
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.48 Petal.Length.48 Petal.Width.48 Species.48 Sepal.Length.49
#> 1            3.5             1.4            0.2     setosa             5.1
#>   Sepal.Width.49 Petal.Length.49 Petal.Width.49 Species.49
#> 1            3.5             1.4            0.2     setosa
iris[1, rep_len(seq_len(ncol(iris)), 251)]
#>      Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>      Sepal.Length.1 Sepal.Width.1 Petal.Length.1 Petal.Width.1 Species.1
#>      Sepal.Length.2 Sepal.Width.2 Petal.Length.2 Petal.Width.2 Species.2
#>      Sepal.Length.3 Sepal.Width.3 Petal.Length.3 Petal.Width.3 Species.3
#>      Sepal.Length.4 Sepal.Width.4 Petal.Length.4 Petal.Width.4 Species.4
#>      Sepal.Length.5 Sepal.Width.5 Petal.Length.5 Petal.Width.5 Species.5
#>      Sepal.Length.6 Sepal.Width.6 Petal.Length.6 Petal.Width.6 Species.6
#>      Sepal.Length.7 Sepal.Width.7 Petal.Length.7 Petal.Width.7 Species.7
#>      Sepal.Length.8 Sepal.Width.8 Petal.Length.8 Petal.Width.8 Species.8
#>      Sepal.Length.9 Sepal.Width.9 Petal.Length.9 Petal.Width.9 Species.9
#>      Sepal.Length.10 Sepal.Width.10 Petal.Length.10 Petal.Width.10
#>      Species.10 Sepal.Length.11 Sepal.Width.11 Petal.Length.11
#>      Petal.Width.11 Species.11 Sepal.Length.12 Sepal.Width.12
#>      Petal.Length.12 Petal.Width.12 Species.12 Sepal.Length.13
#>      Sepal.Width.13 Petal.Length.13 Petal.Width.13 Species.13
#>      Sepal.Length.14 Sepal.Width.14 Petal.Length.14 Petal.Width.14
#>      Species.14 Sepal.Length.15 Sepal.Width.15 Petal.Length.15
#>      Petal.Width.15 Species.15 Sepal.Length.16 Sepal.Width.16
#>      Petal.Length.16 Petal.Width.16 Species.16 Sepal.Length.17
#>      Sepal.Width.17 Petal.Length.17 Petal.Width.17 Species.17
#>      Sepal.Length.18 Sepal.Width.18 Petal.Length.18 Petal.Width.18
#>      Species.18 Sepal.Length.19 Sepal.Width.19 Petal.Length.19
#>      Petal.Width.19 Species.19 Sepal.Length.20 Sepal.Width.20
#>      Petal.Length.20 Petal.Width.20 Species.20 Sepal.Length.21
#>      Sepal.Width.21 Petal.Length.21 Petal.Width.21 Species.21
#>      Sepal.Length.22 Sepal.Width.22 Petal.Length.22 Petal.Width.22
#>      Species.22 Sepal.Length.23 Sepal.Width.23 Petal.Length.23
#>      Petal.Width.23 Species.23 Sepal.Length.24 Sepal.Width.24
#>      Petal.Length.24 Petal.Width.24 Species.24 Sepal.Length.25
#>      Sepal.Width.25 Petal.Length.25 Petal.Width.25 Species.25
#>      Sepal.Length.26 Sepal.Width.26 Petal.Length.26 Petal.Width.26
#>      Species.26 Sepal.Length.27 Sepal.Width.27 Petal.Length.27
#>      Petal.Width.27 Species.27 Sepal.Length.28 Sepal.Width.28
#>      Petal.Length.28 Petal.Width.28 Species.28 Sepal.Length.29
#>      Sepal.Width.29 Petal.Length.29 Petal.Width.29 Species.29
#>      Sepal.Length.30 Sepal.Width.30 Petal.Length.30 Petal.Width.30
#>      Species.30 Sepal.Length.31 Sepal.Width.31 Petal.Length.31
#>      Petal.Width.31 Species.31 Sepal.Length.32 Sepal.Width.32
#>      Petal.Length.32 Petal.Width.32 Species.32 Sepal.Length.33
#>      Sepal.Width.33 Petal.Length.33 Petal.Width.33 Species.33
#>      Sepal.Length.34 Sepal.Width.34 Petal.Length.34 Petal.Width.34
#>      Species.34 Sepal.Length.35 Sepal.Width.35 Petal.Length.35
#>      Petal.Width.35 Species.35 Sepal.Length.36 Sepal.Width.36
#>      Petal.Length.36 Petal.Width.36 Species.36 Sepal.Length.37
#>      Sepal.Width.37 Petal.Length.37 Petal.Width.37 Species.37
#>      Sepal.Length.38 Sepal.Width.38 Petal.Length.38 Petal.Width.38
#>      Species.38 Sepal.Length.39 Sepal.Width.39 Petal.Length.39
#>      Petal.Width.39 Species.39 Sepal.Length.40 Sepal.Width.40
#>      Petal.Length.40 Petal.Width.40 Species.40 Sepal.Length.41
#>      Sepal.Width.41 Petal.Length.41 Petal.Width.41 Species.41
#>      Sepal.Length.42 Sepal.Width.42 Petal.Length.42 Petal.Width.42
#>      Species.42 Sepal.Length.43 Sepal.Width.43 Petal.Length.43
#>      Petal.Width.43 Species.43 Sepal.Length.44 Sepal.Width.44
#>      Petal.Length.44 Petal.Width.44 Species.44 Sepal.Length.45
#>      Sepal.Width.45 Petal.Length.45 Petal.Width.45 Species.45
#>      Sepal.Length.46 Sepal.Width.46 Petal.Length.46 Petal.Width.46
#>      Species.46 Sepal.Length.47 Sepal.Width.47 Petal.Length.47
#>      Petal.Width.47 Species.47 Sepal.Length.48 Sepal.Width.48
#>      Petal.Length.48 Petal.Width.48 Species.48 Sepal.Length.49
#>      Sepal.Width.49 Petal.Length.49 Petal.Width.49 Species.49
#>      Sepal.Length.50
#>  [ reached 'max' / getOption("max.print") -- omitted 1 rows ]
```

<sup>Created on 2019-07-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This seems very strange that one column more of `max.print` get us no rows in output. 🤔  

Coupling this observation with the `hms()` class issue get us the warning. 
`lens != nr` in `format.data.frame` for column with `hms` class

Here is a reprex
``` r
x <- readr::type_convert(data.frame(time = "02:03", other = TRUE, num = "10", 
                                    stringsAsFactors = FALSE))[integer(0), , drop = FALSE]
#> Parsed with column specification:
#> cols(
#>   time = col_time(format = ""),
#>   num = col_double()
#> )
rval <- vector("list", length(x))
nc <- length(x)
for(i in seq_len(nc))
  rval[[i]] <- format(x[[i]], justify = "none")
rval
#> [[1]]
#> [1] "hms()"
#> 
#> [[2]]
#> character(0)
#> 
#> [[3]]
#> character(0)
lens <- vapply(rval, NROW, 1)
lens
#> [1] 1 0 0
nr <- .row_names_info(x, 2L)
nr
#> [1] 0
lens != nr
#> [1]  TRUE FALSE FALSE
```

<sup>Created on 2019-07-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

So there is something off with `hms` class when 0 element.

OK, this is all for now. 

@jimhester I really would like your thought on this one. I think there could be a bug in R source but I don't even know how to report or ask. Also I did not mange to find easily in which version the code was modified. I'll try later.

And about the reported bug,  I think this is in `hms` more than `readr`

**Comment by jimhester** [comment]
I don't think there is a bug in R. `max.print` is the number of _entries_ to limit, so if it is less than the number of columns not all the columns will be printed. 

However the issue you opened in hms seems a good one, `format.hms()` should probably return `character(0)` for an empty value, to avoid this issue. You can actually avoid the original error by setting a higher value of the `max.print` option, e.g. `options(max.print = 99999)` then printing the data.frame works fine.

In any case this is not a bug in readr, the error is coming from formatting the output and the hms package, so I will close this issue.

Thank you for the investigation @cderv!

**Comment by cderv** [comment]
> I don't think there is a bug in R. max.print is the number of entries to limit, so if it is less than the number of columns not all the columns will be printed.

Ok I see. :sweat_smile: So this is not lines but entries as in number of "cells" for a data.frame. With this is mind, the result is clearer ! 
I completety misunderstood this :confused: - for print data.frame I was expecting lines.  Thanks for the clarification ! Always nice to learn new things. :relaxed:
